cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -213,12 +205,15 @@ protected void shutDown() throws Exception {
                                          com.google.common.base.Predicate<ApplicationRecord> predicate)
     throws Exception {
     List<ApplicationRecord> appRecords = new ArrayList<>();
-    Set<ApplicationId> appIds = new HashSet<>();
+    Map<ApplicationId, ApplicationSpecification> appSpecs = new HashMap<>();
     for (ApplicationSpecification appSpec : store.getAllApplications(namespace)) {
-      appIds.add(namespace.app(appSpec.getName(), appSpec.getAppVersion()));
+      appSpecs.put(namespace.app(appSpec.getName(), appSpec.getAppVersion()), appSpec);
     }
+    appSpecs.keySet().retainAll(authorizationEnforcer.isVisible(appSpecs.keySet(),
+                                                                authenticationContext.getPrincipal()));
 
-    for (ApplicationId appId : appIds) {
+    for (EntityId entityId : appSpecs.keySet()) {
+      ApplicationId appId = (ApplicationId) entityId;
       ApplicationSpecification appSpec = store.getApplication(appId);
```

Comment:
The spec can be got by looking up `Map<ApplicationId, ApplicationSpecification> appSpecs`, right?